### PR TITLE
making more corrections [batch #2] to the P4₁₆ spec. based on reading version 1.2.2

### PR DESCRIPTION
**Pull request by Abe149**
Making more corrections to the P4₁₆ spec. based on reading version 1.2.2 [batch number 2].

### CODE REVIEW

File: p4-16/spec/P4-16-spec.mdk [hunk 1]
```diff
@@ -3732,39 +3732,39 @@ finding this information in a section dedicated to type `varbit`.
...
-for programmers, it has several benefits:
+permitted on base types and derived types introduced by `typedef`, `type`, and `enum`.
+While this design is arguably more onerous for programmers, it has several benefits:
 
 - It makes user intent unambiguous.
 - It makes the costs associated with converting numeric values
-  explicit. Implementing certain casts involve sign-extensions, and
-  thus can require significant computational resources on some
-  targets.
+  explicit. Implementing certain casts involve sign extensions,
```
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk:3740]: Why not "involves"?

File: p4-16/spec/P4-16-spec.mdk [hunk 2]
```diff
@@ -3732,39 +3732,39 @@ finding this information in a section dedicated to type `varbit`.
...
   specification. Some targets may not support all casts.
 
 ### Explicit casts
 
 The following casts are legal in P4:
 
-- `bit<1> <-> bool`: converts the value `0` to `false`, the value `1` to `true`, and vice versa.
-- `int -> bool`: only if the `int` value is `0` (converted to `false`) or `1` (converted to `true`)
-- `int<W> -> bit<W>`: preserves all bits unchanged and reinterprets negative
+- `bit<1> ↔ bool`: converts the value `0` to `false`, the value `1` to `true`, and vice versa.
```
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk:3749]: This arrow is unreadable at least in the markup file. Also, latex is not necessarily Unicode-friendly, did you check to see how it looks in the pdf?

File: p4-16/spec/P4-16-spec.mdk [hunk 3]
```diff
@@ -4156,9 +4156,9 @@ of the header.
 "false". It can only be applied to an l-value.
 
 Similar to a `struct`, a header object can be initialized with a list
-expression [#sec-list-exprs] --- the list fields are assigned to the
+expression (see section [#sec-list-exprs]) --- the list fields are assigned to the
```
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk:4159]: Isn't "Section" supposed to be capitalized as well?

File: p4-16/spec/P4-16-spec.mdk [hunk 4]
```diff
@@ -4210,11 +4210,11 @@ Given a header stack value `hs` of size `n`, the following
 expressions are legal:
 
 - `hs[index]`: produces a reference to the header at the
-  specified position within the stack; if `hs` is an l-value, the
-  result is also an l-value. The header may be invalid. Some implementations
+  specified position within the stack; if `hs` is an l-value,
+  the result is also an l-value. The header may be invalid. Some implementations
```
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk:4214]: I would appreciate refraining from making just space changes, they add to the review burden significantly.

File: p4-16/spec/P4-16-spec.mdk [hunk 5]
```diff
@@ -4210,11 +4210,11 @@ Given a header stack value `hs` of size `n`, the following
...
   may impose the constraint that the index expression evaluates to a
-  compile-time known value. A P4 compiler must give an error if an
-  index value that is a compile-time constant is out of range.
+  value that is known at compile time. A P4 compiler must give an error
```
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk:4216]: Actually this is a term that we tried to use consistently through the document: "compile-time known value".

**Review by mbudiu-vmw — COMMENTED** [review]
I will not continue to review this document. We had debates in the working group about the meaning of "compile-time known values" which are not settled. These changes cannot be approved.

If you would submit smaller PRs we could perhaps take them individually.

File: p4-16/spec/P4-16-spec.mdk (hunk 2 above)
Comment (Abe149) [reply p4-16/spec/P4-16-spec.mdk:3749]: AFAIK this spec is not written in the LᴀTᴇX language, but rather in the “Madoko” language.

Comment (mbudiu-vmw) [reply p4-16/spec/P4-16-spec.mdk:3749]: Yes, but it produces two consumables: one in pdf, through latex, and one in html.
The question is about how these look.

File: p4-16/spec/P4-16-spec.mdk (hunk 5 above)
Comment (Abe149) [reply p4-16/spec/P4-16-spec.mdk:4216]: Yes, but it`s poor English.

I`m guessing nobody other than myself would use “{compilation time}-known value” or “compileTime-known value”, so I tried to come up with a semantically-equivalent phrase that is good English.

**Comment by Abe149** [comment]
> If you would submit smaller PRs we could perhaps take them individually.

Understood.

File: p4-16/spec/P4-16-spec.mdk (hunk 3 above)
Comment (Abe149) [reply p4-16/spec/P4-16-spec.mdk:4159]: Fine with me.

Comment (Abe149) [reply p4-16/spec/P4-16-spec.mdk]: https://github.com/p4lang/p4-spec/pull/1081/commits/43f0d41e6a791c682ff802771f731346ca770cc7

File: p4-16/spec/P4-16-spec.mdk (hunk 1 above)
Comment (Abe149) [reply p4-16/spec/P4-16-spec.mdk:3740]: A pre-Abe bug I did _not_ catch right next to a bug I _did_ catch.  Thanks.  Adding a fix in my private repo. (should be included in my next PR).

File: p4-16/spec/P4-16-spec.mdk (hunk 2 above)
Comment (Abe149) [reply p4-16/spec/P4-16-spec.mdk:3749]: After some testing [using <https://hub.docker.com/r/p4lang/madoko-debian>], I found that Unicode support was broken by default.

After extensive experimentation, I found that only 1 out of 4 different ways of trying to fix it actually _works_.

image

image

Comment (Abe149) [reply p4-16/spec/P4-16-spec.mdk]: ### Before (from r1.2.2)
image

### After (from my latest WIP)
image

**Pull request closed without merging**